Allow setting max forks and max concurrent jobs per container/instance group

kdelee opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

Implements https://github.com/ansible/awx/issues/12288, which asked for "max jobs/container group" but in a more generic way, allowing users to set max_concurrent_jobs or max_forks, or both on any instance group.

For container groups, max_concurrent_jobs makes sense when you know what the pod resource requests are and what the global resource limit is on the kubernetes namespace, and you want a global limit of how many of these pods you will provision. However, it does not take into account the variability of how many forks and individual job will spawn of ansible. Allowing the setting of max_forks as well allows scenarios like:

max_concurrent_jobs: 10
max_forks: 50

If 10 jobs that use 5 forks each are run, an 11th job will wait until one of these finishes to run on that group (or be scheduled on a different group with capacity).
If 2 jobs are running with 20 forks each, then a 3rd job with a task_impact of 11 or more will wait until one of these finishes to run on that group (or be scheduled on a different group with capacity).

This does not implement the idea of dynamic resource requests on the pod as described in https://github.com/ansible/awx/issues/12019 -- but it DOES put us in the position where we can say "oh, max_forks on this container group is 100 and I have 5 jobs with forks set to 20, stop scheduling pods", which may be good enough! Especially if pod specs only set requests and not limits, the pods can hypothetically take as many resources as they want (as long as they don't surpass other more global limits that would cause kube to come reap the pod)


##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New or Enhanced Feature

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### Additional Questions/Followup work

The next question is if we would want to be able to set this up from Day 0 with settings in the operator, at least for the default container group. That is probably feasible. But it gets me thinking, with all the work going into things like the awx.awx collection and the CasC collection https://github.com/redhat-cop/controller_configuration , if we instead of adding more an more options for that type of config to settings/the operator etc, if we want to have some other kind of first class way of having a post-install hook that can apply CasC config. :zap:

kdelee committed:
Add max concurrent jobs and max forks per ig

The intention of this feature is primarily to provide some notion of max
capacity of container groups, but the logic I've left generic. Default
is 0, which will be interpereted as no maximum number of jobs or forks.

Includes refactor of variable and method names for clarity.
instances_by_hostname is an internal attribute of TaskManagerInstances.
Clarify when we are expecting the actual TaskManagerInstances object.

Unify how we process running tasks and consume capacity. This has the
effect that we do less expensive work in after_lock_init and have 1 less
loop over all the running tasks. Previously we looped for both building
the dependency graph as well as for calculating the starting capacity of
all the instances and instance groups. Now we acheive both tasks in the
same loop.

Because of how this changes the somewhat subtle "do-si-do" of how to
initialize the Task Manager models, introduce a wrapper class that tries
to take some of that burden off of other areas where we re-use this like
in the serializer and the metrics. Also use this wrapper class to handle
nicities of how to track capacity consumption on instances and instance
groups.

Add tests for max_forks and max_concurrent_jobs

Fixup tests that use TaskManagerModels to accomodate changes.

assign ig before call to consume capacity

if we don't do it in that order, then we don't correctly account for
the container group jobs we are starting in the middle of the task
manager run

kdelee commented on awx/main/scheduler/task_manager_models.py:
```diff
@@ -33,9 +33,55 @@ def remaining_capacity(self):
+            return
+        self.running_jobs += 1
+        if task and (task.execution_node in self.instance_hostnames or self.is_container_group):
+            self.consumed_forks += task.task_impact
```
Ok, just remembered why this never worked in the old task manager, which tried to track this (but for no reason)

Instances can be members of multiple instance groups

This implementation is naive, we consume capacity when the instance group is assigned.
For container groups, this will always work, since they don't have instances.
But for regular instance groups, what about when an instance is in multiple instance groups? Do we have a reverse mapping of `instance: list of instance groups` and consume capacity on all of them? Suppose I could do that

kdelee replied:
Resolved this now

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -626,15 +625,21 @@ def process_pending_tasks(self, pending_tasks):
             # All task.capacity_type == 'control' jobs should run on control plane, no need to loop over instance groups
             if task.capacity_type == 'control':
+                if not self.instance_groups[self.controlplane_ig.name].has_remaining_capacity(control_impact=True):
+                    logger.debug(f"Skipping instance group {self.controlplane_ig.name}, already met max forks allowed to execute on Instance Group")
```
This log statement could use some added context about `task`, otherwise it'll be a bunch of repeated and indecipherable logs.

AlanCoding commented on awx/main/scheduler/task_manager_models.py:
```diff
@@ -64,58 +133,67 @@ def __contains__(self, hostname):
+    def consume_capacity(self, task):
+        self.task_manager_instances.consume_capacity(task)
+        if task.instance_group is not None and task.instance_group.is_container_group:
+            self.instance_groups[task.instance_group.name].consume_capacity(task)
```
```python
        if task.instance_group_id:
             ig = self.pk_ig_map[task.instance_group_id]
            if ig.is_container_group:
                self.instance_groups[ig.name].consume_capacity(task)
```

for performance, I think this will allow dropping the prefetch of `instance_group`

AlanCoding commented on awx/main/scheduler/task_manager_models.py:
```diff
@@ -139,3 +228,20 @@ def get_instance_groups_from_task_cache(self, task):
+
+                tasks = UnifiedJob.objects.filter(status__in=['running', 'waiting']).only('task_impact', 'controller_node', 'execution_node', 'instance_group')
+            for task in tasks:
+                self.instance_groups.consume_capacity(task)
```
I would like to better isolate `init_consumed_capacity` from the logic that runs in the task manager by a separate `@classmethod` constructor.

like

```python
@classmethod
def init_with_consumed_capacity(cls, **kwargs):
    tmm = cls(**kwargs)
    tasks = kwargs.get('tasks', None)
    if tasks is None:
        from awx.main.models import UnifiedJob

        tasks = UnifiedJob.objects.filter(status__in=['running', 'waiting']).only('task_impact', 'controller_node', 'execution_node', 'instance_group')
    for task in tasks:
        self.instance_groups.consume_capacity(task)
```

Then call by `TaskManagerModels.init_with_consumed_capacity()`

kdelee replied:
sounds good, will do

kdelee replied:
done

AlanCoding commented on awx/main/scheduler/task_manager_models.py:
```diff
@@ -64,61 +137,77 @@ def __contains__(self, hostname):
+        if task.instance_group_id:
+            ig = self.pk_ig_map[task.instance_group_id]
+            if ig.is_container_group:
+                self.instance_groups[ig.name].consume_capacity(task)
```
I find it a little unexpected that `TaskManagerInstanceGroups.consume_capacity` deducts the capacity from the instance group as well as the instance(s) the job runs on.

There are 4 methods for `consume_capacity`, and this isn't a bad thing, but with that number it becomes hard to predict, so it might be best to follow expectations as closely as possible. As it stands, it might be more consistent to introduce a `TaskManagerModels.consume_capacity` method to deduct capacity from both the IGs and the instances, moving the `self.task_manager_instances.consume_capacity(task)` line from here to there. Maybe ask @fosterseth for his input on this matter as well. My suggestion here would require changing the task manager to keep the `TaskManagerModels` instance as opposed to the `TaskManagerInstanceGroups`, but that might obligate some other changes to the references between classes.

kdelee replied:
I agree its a bit weird to have the method on `TaskManagerInstanceGroups` but have it apply to groups and instances, I'll shop some ideas around with seth and we can move forward. I don't think any other place other than the task manager ever uses any `consume_capacity` function so it should be relatively small lift

kdelee replied:
implemented the suggestion

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -626,15 +629,19 @@ def process_pending_tasks(self, pending_tasks):
 
             for instance_group in self.instance_groups.get_instance_groups_from_task_cache(task):
+                if not self.instance_groups[instance_group.name].has_remaining_capacity(task):
+                    continue
```
I do think we should have a log, like:

```python
                    logger.debug(
                        f"Instance group {instance_group.name} not available to run job {task.log_format} w/ capacity requirement {task.task_impact}"
                    )
```

kdelee replied:
I did add the log in the `has_remaining_capacity` though probably should update to use `task.log_format` i'll comment there to point it out

kdelee replied:
pointed out later where I do log in the case that we max_forks or max_concurrent_jobs is blocking (only at debug level)

AlanCoding commented:
This looks good to me from the API code design side of things :+1: 

There are a few more boxes to check, like testing and UI, but the main API development looks good.

kdelee commented on awx/main/scheduler/task_manager_models.py:
```diff
@@ -33,28 +36,98 @@ def remaining_capacity(self):
+                logger.debug(f"{task_string} cannot fit on instance group {self.name} with {remaining_jobs} remaining jobs")
+            if remaining_forks < 0:
+                impact_string = f"with impact {task_impact}"
+                logger.debug(f"{task_string} {impact_string} cannot fit on instance group {self.name} with {remaining_forks} remaining forks")
```
here is where I'm logging if indeed the instance group does NOT have capacity, and it tells you why (either hit max forks or max concurrent jobs)

kdelee committed:
Add max_forks, max_concurrent_jobs to collection

Now these items can be set on instance_groups via the collection.

kdelee committed:
update UI for max_forks and max_concurrent_jobs

This includes updates for container groups and instance groups.
Includes updates to ui test data with new fields

kdelee commented:
![Screenshot from 2022-10-26 22-04-24](https://user-images.githubusercontent.com/29239279/198174196-7cdfc4bb-b218-40d5-be30-edc3a4c07121.png)
![Screenshot from 2022-10-26 22-04-17](https://user-images.githubusercontent.com/29239279/198174197-5da9a451-1bd7-4f8d-90ea-eaaa2bc8e093.png)
![Screenshot from 2022-10-26 22-04-05](https://user-images.githubusercontent.com/29239279/198174199-1e1e0c93-2afc-4127-b190-08008320b1cf.png)
![Screenshot from 2022-10-26 22-03-58](https://user-images.githubusercontent.com/29239279/198174201-39b36862-bac9-49cb-a0d0-8dce3368a431.png)

![image](https://user-images.githubusercontent.com/29239279/198174310-a0510d6b-427d-49f5-8d09-1bfa8decccf1.png)

AlexSCorey reviewed (approved):
Looks good to me.

kdelee committed:
use task manager models more consistently in serializer

AlanCoding commented:
Summarizing remaining actions left to do

### Setting-based default

To make this fit the intended use in OCP, we need to have a setting which will give _the default value for the OCP `default` instance group_. This instance group is created in some bootstrap scripts by the `awx-manage register_queue` command, and we don't want to change that, it will be easier to have it get a default value from a setting.

### Docs

We need some some highly specific instructions for the primary intended use case. I find this very similar to `SYSTEM_TASK_ABS_MEM`, which is used to correct the capacity of control plane instances.

I'd like our docs approach to tackle the general problem of "how do I assure my OCP cluster is using accurate capacity for job distribution?" Then answer would be to add some things to `extra_settings` in the operator deployment where you set that old setting, as well as the new setting.

### Integration testing

It's expected that a serial test will be needed to exercise the basic functionality of this feature.

kdelee committed:
settings for max forks/job on default job queue

This will allow users of the operator to set these settings
so from the start when the operator creates the default
execution queue they can control the max_forks and max_concurrent_jobs
on the default container group.

kdelee commented:
@AlanCoding ok added the settings for the default container group and the API tests/operator tests passing

kdelee committed:
add new fields to awxkit

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -471,9 +475,8 @@ def after_lock_init(self):
-        self.instance_groups = TaskManagerInstanceGroups(instances_by_hostname=self.instances)
-        self.controlplane_ig = self.instance_groups.controlplane_ig
+        self.tm_models = TaskManagerModels()
+        self.controlplane_ig = self.tm_models.instance_groups.controlplane_ig
```
super duper minor comment, but I'd prefer to save `controlplane_ig` as a local variable in process_pending_tasks instead of `self.controlplane_ig`.

But the semantic of `self.tm_models` is an improvement from before, this is going in the right directions.

AlanCoding reviewed (approved):
I had already gone through this pretty completely before. I wanted to make sure the core feature testing was having success, and it is now. :100:

kdelee committed:
filter tasks when instance groups are filtered

this is necessary when requests are made to to
api/v2/job_templates/ID/instance_groups

Thanks to Sarah who found this!

akus062381 commented:
I was running into errors in the UI/API, and Elijah helped me diagnose that the error is related to this endpoint:
api/v2/job_templates/7/instance_groups/

akus062381 commented:
this looks great

kdelee commented:
> I was running into errors in the UI/API, and Elijah helped me diagnose that the error is related to this endpoint: api/v2/job_templates/7/instance_groups/

We fixed this now! Thanks @akus062381

kdelee commented on awx/main/scheduler/task_manager_models.py:
```diff
@@ -139,3 +254,56 @@ def get_instance_groups_from_task_cache(self, task):
+                # we also need to filter the tasks we look at
+                tasks = UnifiedJob.objects.filter(status__in=task_status_filter_list, instance_group__in=[ig.id for ig in instance_group_queryset]).only(
+                    *task_fields
+                )
```
@AlanCoding FYI @akus062381 found a bug in this when we were making requests to /api/v2/job_templates/<ID>/instance_groups, the serializer calls this with a instance group queryset already formed. So you could get a 500 if there were jobs running on a DIFFERENT instance group that wasn't in the queryset, because later in `consume_capacity` we assumed all the instance groups on the tasks should be in the instance group mapping

AlanCoding replied:
Oh, I get what you're saying about this but. Good find.

I'm on the fence about whether `instance_groups_queryset` is worth the effort we're putting in to maintain it. Think about the multiple types of optimizations. We have a single IG in a list view of IGs. The main optimization the serializer is trying to accomplish is to avoid re-computing capacity for every object in the list. Even if we incur 1 additional query, we can save those extra 20.

On the other hand, I see how we avoid tracking capacity for a _large number_ of instance groups if the user created a lot...

kdelee replied:
yeah so alternative to this fix is to drop the instance_group_queryset argumetn and always compute everything for all groups and all tasks......

but there is a tiny part of me that is still dreaming of partitioning the task manager work on instance groups and making it where instances can only belong to 1 instance group :smiling_imp:

matburt commented:
Going into the merge review meeting I'd love to see details about this added to the right files in `docs/` : https://github.com/ansible/awx/tree/devel/docs

AlanCoding commented:
I can see how `docs/capacity.md` is an appropriate context for a summary of this feature. Looking at the doc now, I don't see anything outright wrong, unlike `docs/task_manager_system.md`, which needs substantial refactoring.

kdelee commented:
![image](https://user-images.githubusercontent.com/29239279/204571051-bfba393d-e130-41e8-8db4-b5f0e8b43898.png)
 bug! need to fix

kdelee commented:
@AlexSCorey helped me fix that form edit problem so we can successfully edit and set values back to 0

kdelee committed:
add docs about max_forks and max_concurrent_jobs

Add this to the capacity.md that also dicsusses details about
task_impact and forks and how that is used in the application.
